File: Odd Even Linked List.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        current_odd = head
        if head:
            even_start = head.next
            current_even = head.next
        else:
            return current_odd
        
        while current_odd and current_even:
            current_odd.next = current_even.next
            if current_odd.next:
                current_odd = current_odd.next
            if current_even and current_odd:
                current_even.next = current_odd.next
                current_even = current_even.next

        current_odd.next = even_start
        return head

File: cpp/Odd Even Linked List.cpp
#include "common.h"

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    // This does it by the value of the node and not by 
    // the index being odd or even
    ListNode* oddEvenList_by_value(ListNode* head) {
        ListNode* even_head = nullptr;
        ListNode* odd_head = nullptr;
        ListNode* current_even = even_head;
        ListNode* current_odd = odd_head;
        ListNode* current = head;

        while (current) {
            ListNode* next_node = current->next;
            current->next = nullptr;

            if (current->val % 2) {
                if (current_odd) {
                    current_odd->next = current;
                    current_odd = current;
                } else {
                    odd_head = current;
                    current_odd = odd_head;
                }
            } else {
                if (current_even) {
                    current_even->next = current;
                    current_even = current;
                } else {
                    even_head = current;
                    current_even = even_head;
                }
            }
            current = next_node;
        }
        if (current_odd) {
            current_odd->next = even_head; 
        }
        return odd_head ? odd_head : even_head;
    }

    
    ListNode* oddEvenList(ListNode* head) {
        if (!head) return nullptr;
        ListNode* odd = head;
        ListNode* even = head->next;
        ListNode* evenHead = even;

        while (even && even->next) {
            odd->next = even->next;
            odd = odd->next;
            even->next = odd->next;
            even = even->next;
        }
        odd->next = evenHead;
        return head;
    }


};